File: 0016-3sum-closest/0016-3sum-closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        int diff = Integer.MAX_VALUE;
        int closestSum = -1;
        Arrays.sort(nums);
        
        for(int i=0; i<nums.length-2; i++) {
            int l = i+1;
            int r = nums.length-1;
            
            while(l < r) {
                int sum = nums[i] + nums[l] + nums[r];
                if(sum == target) return sum;
                else {
                    int tempDiff = Math.abs(sum-target);
                    if(tempDiff < diff) {
                        diff = tempDiff;
                        closestSum = sum;
                    }
                    
                    if(sum < target) {
                        l++;
                    } else {
                        r--;
                    }
                }
            }
        }
        return closestSum;
    }
}

File: 0016-3sum-closest/0016-3sum-closest.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function(nums, target) {
    nums.sort((a, b) => a-b)
    let sum = 0
    let diff = Number.MAX_SAFE_INTEGER
    for(let i=0; i<nums.length; i++) {
        const num = nums[i]
        let l = i+1,  r = nums.length-1

        while(l < r) {
            const currsum = num + nums[l] + nums[r]
            if(Math.abs(currsum-target) < diff) {
                diff = Math.abs(currsum-target)
                sum = currsum
            }
            if(currsum < target) {
                l++
            } else {
                r--
            }
        }
    }
    return sum
};